File: backend/src/InsuraTech.Domain/Policies/Policy.cs
using InsuraTech.Domain.Common;
using InsuraTech.Domain.Events;
using InsuraTech.Domain.Exceptions;
using InsuraTech.Domain.Policies.HealthPlan;
using InsuraTech.Domain.Policies.LifePlan;
using InsuraTech.Domain.Policies.TravelPlan;
using InsuraTech.Domain.Policies.ValueObjects;
using InsuraTech.Domain.Policies.HomePlan;
using InsuraTech.Domain.Policies.VehiclePlan;

namespace InsuraTech.Domain.Policies
{
    public sealed class Policy : AggregateRoot
    {
        public PolicyNumber Number { get; private set; } = null!;
        public PolicyType Type { get; private set; }
        public PolicyStatus Status { get; private set; }
        public InsuredPerson Insured { get; private set; } = null!;
        public CoveragePeriod Coverage { get; private set; } = null!;
        public decimal MonthlyPremium { get; private set; }
        public decimal InsuredAmount { get; private set; }
        public decimal AvailableInsuredAmount { get; private set; }
        public string? CancellationReason { get; private set; }
        public DateOnly? CancellationEffectiveDate { get; private set; }
        public Guid? RenewedFromPolicyId { get; private set; }
        public HealthPlanSelection? HealthPlan { get; private set; }
        public TravelPlanSelection? TravelPlan { get; private set; }
        public LifePlanSelection? LifePlan { get; private set; }
        public VehiclePlanSelection? VehiclePlan { get; private set; }
        public HomePlanSelection?    HomePlan    { get; private set; }

        /// <summary>Id del asesor que creó la póliza. Null si fue creada directamente por Admin/Leader.</summary>
        public Guid? CreatedByAdvisorId { get; private set; }

        private readonly List<PolicyStatusHistory> _statusHistory = new();
        public IReadOnlyCollection<PolicyStatusHistory> StatusHistory =>
            _statusHistory.AsReadOnly();

        private Policy() { }

        // Factory — tipos con monto manual (Life transitorio; Vehicle y Home en futuras specs)
        public static Policy Create(
            PolicyNumber number,
            PolicyType type,
            InsuredPerson insured,
            CoveragePeriod coverage,
            decimal monthlyPremium,
            decimal insuredAmount)
        {
            if (insuredAmount <= 0)
                throw new ArgumentException("Insured amount must be greater than zero.", nameof(insuredAmount));
            if (monthlyPremium <= 0)
                throw new ArgumentException("Monthly premium must be greater than zero.", nameof(monthlyPremium));

            var policy = new Policy
            {
                Number = number,
                Type = type,
                Insured = insured,
                Coverage = coverage,
                MonthlyPremium = monthlyPremium,
                InsuredAmount = insuredAmount,
                AvailableInsuredAmount = insuredAmount,
                Status = PolicyStatus.Pending
            };

            policy.AddStatusHistory(PolicyStatus.Pending, "Policy created.");
            return policy;
        }

        // Factory — tipo Health (plan predefinido + cálculo automático)
        public static Policy CreateHealthPolicy(
            PolicyNumber number,
            InsuredPerson insured,
            CoveragePeriod coverage,
            decimal monthlyPremium,
            string healthPlanId,
            DateOnly today)
        {
            var selection = HealthPlanPricingService.Calculate(healthPlanId, insured.BirthDate, today);

            if (monthlyPremium <= 0)
                throw new ArgumentException("Monthly premium must be greater than zero.", nameof(monthlyPremium));

            var policy = new Policy
            {
                Number = number,
                Type = PolicyType.Health,
                Insured = insured,
                Coverage = coverage,
                MonthlyPremium = monthlyPremium,
                InsuredAmount = selection.FinalAmount,
                AvailableInsuredAmount = selection.FinalAmount,
                HealthPlan = selection,
                Status = PolicyStatus.Pending
            };

            policy.AddStatusHistory(PolicyStatus.Pending,
                $"Health policy created with plan '{selection.PlanName}'.");
            return policy;
        }

        // Factory — tipo Life (planes de precio fijo, sin factor de edad, SPEC-009)
        public static Policy CreateLifePolicy(
            PolicyNumber number,
            InsuredPerson insured,
            CoveragePeriod coverage,
            string lifePlanId,
            DateOnly today)
        {
            var selection = LifePlanPricingService.Calculate(lifePlanId, insured.BirthDate, today);

            var policy = new Policy
            {
                Number                 = number,
                Type                   = PolicyType.Life,
                Insured                = insured,
                Coverage               = coverage,
                MonthlyPremium         = selection.MonthlyPremium,
                InsuredAmount          = selection.DeathBenefit,
                AvailableInsuredAmount = selection.DeathBenefit,
                LifePlan               = selection,
                Status                 = PolicyStatus.Pending
            };

            policy.AddStatusHistory(PolicyStatus.Pending,
                $"Life policy created with plan '{selection.PlanName}'.");
            return policy;
        }

        // Factory — tipo Vehicle (motor de cotización por tasa técnica, SPEC-010)
        public static Policy CreateVehiclePolicy(
            PolicyNumber number,
            InsuredPerson insured,
            CoveragePeriod coverage,
            string vehiclePlanId,
            decimal commercialValue,
            int vehicleYear,
            string vehicleBrand,
            DateOnly today)
        {
            var quotation = VehiclePricingService.Calculate(
                commercialValue, vehicleYear, vehicleBrand, today.Year);

            var selection = VehiclePricingService.Select(vehiclePlanId, quotation);

            var policy = new Policy
            {
                Number                 = number,
                Type                   = PolicyType.Vehicle,
                Insured                = insured,
                Coverage               = coverage,
                MonthlyPremium         = selection.FinalMonthlyPremium,
                InsuredAmount          = selection.CommercialValue,
                AvailableInsuredAmount = selection.CommercialValue,
                VehiclePlan            = selection,
                Status                 = PolicyStatus.Pending
            };

            policy.AddStatusHistory(PolicyStatus.Pending,
                $"Vehicle policy created with plan '{selection.PlanName}' for {selection.VehicleBrand} {selection.VehicleYear}.");
            return policy;
        }

        // Factory — tipo Home (motor de cotización por coberturas seleccionadas, SPEC-012)
        public static Policy CreateHomePolicy(
            PolicyNumber                number,
            InsuredPerson               insured,
            CoveragePeriod              coverage,
            string?                     packageId,
            decimal                     propertyValue,
            int                         constructionYear,
            int                         stratum,
            int                         occupants,
            HomePropertyType            propertyType,
            IReadOnlyList<HomeCoverage> selectedCoverages,
            DateOnly                    today)
        {
            var quotation = HomePricingService.Calculate(
                propertyValue, constructionYear, stratum, occupants,
                propertyType, selectedCoverages, today.Year);

            var selection = HomePricingService.Select(packageId, quotation);

            var policy = new Policy
            {
                Number                 = number,
                Type                   = PolicyType.Home,
                Insured                = insured,
                Coverage               = coverage,
                MonthlyPremium         = selection.FinalMonthlyPremium,
                InsuredAmount          = selection.PropertyValue,
                AvailableInsuredAmount = selection.PropertyValue,
                HomePlan               = selection,
                Status                 = PolicyStatus.Pending
            };

            policy.AddStatusHistory(PolicyStatus.Pending,
                $"Home policy created. Package: {selection.PackageName}, " +
                $"Coverages: {selection.SelectedCoverages.Count}, " +
                $"Monthly premium: ${selection.FinalMonthlyPremium:N0} COP.");
            return policy;
        }

        // Factory — tipo Travel (rating engine automático)
        public static Policy CreateTravelPolicy(
            PolicyNumber number,
            InsuredPerson insured,
            CoveragePeriod coverage,
            TripType tripType,
            Continent? continent,
            int durationDays,
            decimal? trmCop,
            DateOnly? trmDate)
        {
            var selection = TravelRatingService.Calculate(
                tripType, continent, durationDays, trmCop, trmDate, DateTime.UtcNow);

            var policy = new Policy
            {
                Number                  = number,
                Type                    = PolicyType.Travel,
                Insured                 = insured,
                Coverage                = coverage,
                MonthlyPremium          = selection.TotalPriceCop,
                InsuredAmount           = selection.TotalPriceCop,
                AvailableInsuredAmount  = selection.TotalPriceCop,
                TravelPlan              = selection,
                Status                  = PolicyStatus.Pending
            };

            policy.AddStatusHistory(PolicyStatus.Pending,
                $"Travel policy created. Type: {tripType}" +
                (continent.HasValue ? $", Continent: {continent}" : string.Empty) +
                $", Duration: {durationDays} days, Total: ${selection.TotalPriceCop:N0} COP.");

            return policy;
        }

        // Poliza Activa
        public void Activate()
        {
            if (Status != PolicyStatus.Pending)
                throw new InvalidPolicyStateException(Status.ToString(), nameof(Activate));

            Status = PolicyStatus.Active;
            MarkAsUpdated();
            IncrementVersion();

            AddStatusHistory(PolicyStatus.Active, "Payment confirmed. Policy activated.");
            AddDomainEvent(new PolicyActivatedEvent(
                Id, Number.Value, Insured.FirstName, Insured.DocumentId));
        }
        // Poliza suspendida

        public void Suspend(string reason)
        {
            if (Status != PolicyStatus.Active)
                throw new InvalidPolicyStateException(Status.ToString(), nameof(Suspend));
            if (string.IsNullOrWhiteSpace(reason))
                throw new ArgumentException("Suspenson reason is requiered. ", nameof(reason));
            Status = PolicyStatus.Suspended;
            MarkAsUpdated();
            IncrementVersion();
            AddStatusHistory(PolicyStatus.Suspended, reason);
        }

        //Poliza cancelada 
        public void Cancel(string reason, DateOnly effectiveDate)
        {
            if (Status == PolicyStatus.Cancelled)
                throw new InvalidPolicyStateException(Status.ToString(), nameof(Cancel));

            if (string.IsNullOrWhiteSpace(reason))
                throw new ArgumentException("Cancellation reason is required.", nameof(reason));

            if (effectiveDate < DateOnly.FromDateTime(DateTime.UtcNow))
                throw new ArgumentException("Effective date cannot be in the past.", nameof(effectiveDate));

            Status = PolicyStatus.Cancelled;
            CancellationReason = reason;
            CancellationEffectiveDate = effectiveDate;
            MarkAsUpdated();
            IncrementVersion();

            AddStatusHistory(PolicyStatus.Cancelled, $"Cancelled: {reason}");
            AddDomainEvent(new PolicyCancelledEvent(Id, Number.Value, reason, effectiveDate));

        }
        // Poliza renovacion

        public Policy Renew(PolicyNumber newNumber, CoveragePeriod newCoverage)
        {
            if (Status != PolicyStatus.Active && Status != PolicyStatus.Expired)
                throw new InvalidPolicyStateException(Status.ToString(), nameof(Renew));

            var renewal = new Policy
            {
                Number = newNumber,
                Type = Type,
                Insured = Insured,
                Coverage = newCoverage,
                MonthlyPremium = MonthlyPremium,
                InsuredAmount = InsuredAmount,
                AvailableInsuredAmount = InsuredAmount,
                Status = PolicyStatus.Pending,
                RenewedFromPolicyId = Id
            };

            renewal.AddStatusHistory(PolicyStatus.Pending, $"Renewed from policy {Number.Value}.");
            return renewal;
        }
        // Monto asegurado deducible cuando hay siniestros

        public void DeductInsuredAmount(decimal amount)
        {
            if (amount <= 0)
                throw new ArgumentException("Amount to deduct must be greater than zero.", nameof(amount));
            if (amount > AvailableInsuredAmount)
                throw new BusinessRuleException("INSUFFICIENT_COVERAGE",
                    $"Claim amount ${amount} exceeds available insured amount ${AvailableInsuredAmount}.");
            AvailableInsuredAmount -= amount;
            MarkAsUpdated();
            IncrementVersion();

            if(AvailableInsuredAmount == 0)
            {
                Status = PolicyStatus.Exhausted;
                AddStatusHistory(PolicyStatus.Exhausted, "Insured amount fully exhausted by approved claims.");
            }
        }

        public bool IsActiveOn(DateOnly date) =>
            Status == PolicyStatus.Active && Coverage.IsActive(date);

        /// <summary>Registra el asesor que creó la póliza. Se invoca desde el handler tras la creación.</summary>
        public void SetCreatedByAdvisor(Guid? advisorId) => CreatedByAdvisorId = advisorId;

        private void AddStatusHistory(PolicyStatus status, string notes) =>
                _statusHistory.Add(PolicyStatusHistory.Create(Id, status, notes));
    }
}

File: backend_v2/src/main/java/com/insuratech/domain/policies/Policy.java
package com.insuratech.domain.policies;

import com.insuratech.domain.common.AggregateRoot;
import com.insuratech.domain.common.PolicyStatusHistory;
import com.insuratech.domain.events.PolicyActivatedEvent;
import com.insuratech.domain.events.PolicyCancelledEvent;
import com.insuratech.domain.events.PolicyExpiringSoonEvent;
import com.insuratech.domain.exceptions.*;
import com.insuratech.domain.policies.health.HealthPlanSelection;
import com.insuratech.domain.policies.home.HomePlanSelection;
import com.insuratech.domain.policies.life.LifePlanSelection;
import com.insuratech.domain.policies.travel.TravelPlanSelection;
import com.insuratech.domain.policies.vehicle.VehiclePlanSelection;
import com.insuratech.domain.policies.vo.CoveragePeriod;
import com.insuratech.domain.policies.vo.InsuredPerson;
import com.insuratech.domain.policies.vo.PolicyNumber;
import lombok.Getter;
import org.springframework.data.mongodb.core.mapping.Document;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

@Getter
@Document(collection = "policies")
public class Policy extends AggregateRoot {

    private PolicyNumber policyNumber;
    private PolicyType policyType;
    private PolicyStatus status;
    private InsuredPerson insuredPerson;
    private CoveragePeriod coveragePeriod;
    private BigDecimal premium;
    private BigDecimal insuredAmount;
    private BigDecimal remainingInsuredAmount;
    private String agentId;
    private String cityId;
    private String notes;

    private HealthPlanSelection healthPlanSelection;
    private LifePlanSelection lifePlanSelection;
    private VehiclePlanSelection vehiclePlanSelection;
    private HomePlanSelection homePlanSelection;
    private TravelPlanSelection travelPlanSelection;

    private final List<PolicyStatusHistory> statusHistory = new ArrayList<>();

    protected Policy() { super(); }

    public static Policy create(PolicyNumber policyNumber, PolicyType policyType,
                                InsuredPerson insuredPerson, CoveragePeriod coveragePeriod,
                                BigDecimal premium, BigDecimal insuredAmount,
                                String agentId, String cityId) {
        Policy policy = new Policy();
        policy.policyNumber = policyNumber;
        policy.policyType = policyType;
        policy.status = PolicyStatus.DRAFT;
        policy.insuredPerson = insuredPerson;
        policy.coveragePeriod = coveragePeriod;
        policy.premium = premium;
        policy.insuredAmount = insuredAmount;
        policy.remainingInsuredAmount = insuredAmount;
        policy.agentId = agentId;
        policy.cityId = cityId;
        return policy;
    }

    public static Policy createHealthPolicy(PolicyNumber policyNumber, InsuredPerson insuredPerson,
                                            CoveragePeriod coveragePeriod, HealthPlanSelection selection,
                                            String agentId, String cityId) {
        Policy policy = create(policyNumber, PolicyType.HEALTH, insuredPerson, coveragePeriod,
            selection.getCalculatedPremium(), selection.getPlan().getCoverageLimit(), agentId, cityId);
        policy.healthPlanSelection = selection;
        return policy;
    }

    public static Policy createLifePolicy(PolicyNumber policyNumber, InsuredPerson insuredPerson,
                                          CoveragePeriod coveragePeriod, LifePlanSelection selection,
                                          String agentId, String cityId) {
        Policy policy = create(policyNumber, PolicyType.LIFE, insuredPerson, coveragePeriod,
            selection.getCalculatedPremium(), selection.getPlan().getCoverageAmount(), agentId, cityId);
        policy.lifePlanSelection = selection;
        return policy;
    }

    public static Policy createVehiclePolicy(PolicyNumber policyNumber, InsuredPerson insuredPerson,
                                             CoveragePeriod coveragePeriod, VehiclePlanSelection selection,
                                             String agentId, String cityId) {
        Policy policy = create(policyNumber, PolicyType.VEHICLE, insuredPerson, coveragePeriod,
            selection.getCalculatedPremium(), selection.getQuotation().getVehicleValue(), agentId, cityId);
        policy.vehiclePlanSelection = selection;
        return policy;
    }

    public static Policy createHomePolicy(PolicyNumber policyNumber, InsuredPerson insuredPerson,
                                          CoveragePeriod coveragePeriod, HomePlanSelection selection,
                                          String agentId, String cityId) {
        Policy policy = create(policyNumber, PolicyType.HOME, insuredPerson, coveragePeriod,
            selection.getCalculatedPremium(), selection.getQuotation().getPropertyValue(), agentId, cityId);
        policy.homePlanSelection = selection;
        return policy;
    }

    public static Policy createTravelPolicy(PolicyNumber policyNumber, InsuredPerson insuredPerson,
                                            CoveragePeriod coveragePeriod, TravelPlanSelection selection,
                                            String agentId, String cityId) {
        Policy policy = create(policyNumber, PolicyType.TRAVEL, insuredPerson, coveragePeriod,
            selection.getCalculatedPremium(), BigDecimal.ZERO, agentId, cityId);
        policy.travelPlanSelection = selection;
        return policy;
    }

    public void activate(String activatedBy) {
        if (status == PolicyStatus.ACTIVE) throw new PolicyAlreadyActiveException();
        status = PolicyStatus.ACTIVE;
        statusHistory.add(new PolicyStatusHistory(PolicyStatus.ACTIVE.name(), activatedBy, "Policy activated"));
        markAsUpdated();
        addDomainEvent(new PolicyActivatedEvent(getId(), policyNumber.getValue()));
    }

    public void suspend(String suspendedBy, String reason) {
        if (status != PolicyStatus.ACTIVE) throw new PolicyNotActiveException();
        status = PolicyStatus.SUSPENDED;
        statusHistory.add(new PolicyStatusHistory(PolicyStatus.SUSPENDED.name(), suspendedBy, reason));
        markAsUpdated();
    }

    public void cancel(String cancelledBy, String reason) {
        if (status == PolicyStatus.CANCELLED)
            throw new BusinessRuleException("PolicyAlreadyCancelled", "Policy is already cancelled.");
        status = PolicyStatus.CANCELLED;
        statusHistory.add(new PolicyStatusHistory(PolicyStatus.CANCELLED.name(), cancelledBy, reason));
        markAsUpdated();
        addDomainEvent(new PolicyCancelledEvent(getId(), reason));
    }

    public void renew(LocalDate newEndDate, String renewedBy) {
        if (status != PolicyStatus.ACTIVE && status != PolicyStatus.EXPIRED)
            throw new BusinessRuleException("PolicyNotRenewable", "Only active or expired policies can be renewed.");
        coveragePeriod = CoveragePeriod.of(coveragePeriod.getEndDate().plusDays(1), newEndDate);
        status = PolicyStatus.ACTIVE;
        statusHistory.add(new PolicyStatusHistory(PolicyStatus.ACTIVE.name(), renewedBy, "Policy renewed"));
        markAsUpdated();
    }

    public void deductInsuredAmount(BigDecimal amount) {
        if (status != PolicyStatus.ACTIVE) throw new PolicyNotActiveException();
        if (amount.compareTo(remainingInsuredAmount) > 0)
            throw new InsufficientInsuredAmountException(amount, remainingInsuredAmount);
        remainingInsuredAmount = remainingInsuredAmount.subtract(amount);
        markAsUpdated();
    }

    public void checkAndRaiseExpiringEvent(int daysThreshold) {
        if (status == PolicyStatus.ACTIVE && coveragePeriod.isExpiringSoon(daysThreshold)) {
            addDomainEvent(new PolicyExpiringSoonEvent(
                getId(), policyNumber.getValue(), coveragePeriod.getEndDate()));
        }
    }

    public List<PolicyStatusHistory> getStatusHistory() {
        return Collections.unmodifiableList(statusHistory);
    }
}
